website/public_adventures.py

```diff
@@ -31,67 +31,166 @@ def __init__(self, db: Database, achievements: Achievements):
-                "levels": adventure.get("levels"),
-                "language": adv_lang,
-                "cloned_times": adventure.get("cloned_times"),
-                "tags": adv_tags,
-                "text": content,
-                "is_teacher_adventure": True,
-            }
-
-            # save adventures for later usage.
-            for _level in adventure.get("levels", [adventure.get("level")]):
-                _level = int(_level)
-                if self.adventures.get(_level):
-                    self.adventures[_level].append(current_adventure)
-                else:
-                    self.adventures[_level] = [current_adventure]
-
-            available_levels = adventure["levels"] if adventure.get("levels") else [adventure["level"]]
-            self.customizations["available_levels"].update([int(adv_level) for adv_level in available_levels])
-
     @route("/", methods=["GET"])
     @route("/filter", methods=["POST"])
     @requires_teacher
-    def filtering(self, user, index_page=False):
+    def filtering(self, user):
+        level = request.args["level"] if request.args.get("level") else "1"
+        language = request.args.get("lang")
+        if language == "reset":
+            language = ""
+        elif not language and request.method == 'GET':
+            language = g.lang
+        tag = request.args.get("tag", "")
+        search = request.form.get("search", request.args.get("search", ""))
+
+        available_languages = set()
+        available_levels = set()
+        available_tags = set()
+
+        # Get all possible filters
+        field_filters = self.db.get_public_adventures_filters()
+
+        available_levels.update(field_filters.get("level", []))
+        available_languages.update(field_filters.get("lang", []))
+        available_tags.update(field_filters.get("tag", []))
+
+        customizations = {"available_levels": sorted(available_levels, key=lambda x: int(x))}
+
+        tags = []
+        if tag:
+            toReset = request.args.get("reset")
+            if toReset:
+                # then it's the current selected tags, so remove current's tag
+                tags = [_tag for _tag in toReset.split(",") if _tag != tag]
+            else:
+                tags = tag.split(",")
+            tags = [t for t in tags if t]
+
+        # Get indexes
+        level_adventure_ids = self.db.get_public_adventures_indexes({"field_value": f"level_{level}"})
+        lang_adventure_ids = self.db.get_public_adventures_indexes({"field_value": f"lang_{language}"})
+        tag_adventure_ids = set()
+        for _t in tags:
+            tag_adventure_ids.update(self.db.get_public_adventures_indexes({"field_value": f"tag_{_t}"}))
+
+        lang_adventure_ids.intersection_update(level_adventure_ids)
+        tag_adventure_ids.intersection_update(level_adventure_ids)
+        adventure_ids = level_adventure_ids.union(lang_adventure_ids, tag_adventure_ids)
```

Comment:
I didn't know `union` takes 3 arguments. What does that mean?

Why are you doing a union here? Aren't we supposed to intersect all 3 filters?

Reply:
Also, where are you accounting for the possibilty that the filter can be empty? What if the user didn't pick a language, level, or set of tags?

Reply:
> Also, where are you accounting for the possibilty that the filter can be empty? What if the user didn't pick a language, level, or set of tags?

Currently, the levels are a must and can't be ignored. If you make an ajax call, it'd be defaulted to level 1.

Reply:
> I didn't know `union` takes 3 arguments. What does that mean?
> 
> Why are you doing a union here? Aren't we supposed to intersect all 3 filters?

I intersect them first then take the union. Because if one is an empty set, the intersection would result in an empty set.

Reply:
I understand, but if both language and tags are given, your set is now:

`adventure_ids` is the set of adventures that have EITHER the given language, OR the given tags.

Is that what you meant? I thought it would have been:

`adventure_ids` is the set of adventures that have BOTH the given language, AND the given tags.

Reply:
> Currently, the levels are a must and can't be ignored. If you make an ajax call, it'd be defaulted to level 1.

That seems like a bummer though? If I just want to see all Spanish adventures, there's no way to do that?

Reply:
> That seems like a bummer though? If I just want to see all Spanish adventures, there's no way to do that?

The main reason behind it is that the js needs to be initialized with a specific level in order to correctly run code in the editor!

Reply:
> `adventure_ids` is the set of adventures that have EITHER the given language, OR the given tags.

What OR? the union will combine these sets right. So, if i have:
```
lang_ids = {'3f8aea42eb324f08a16776671498dd1b'}                                                                                                                                                                       
tag_ids = set()
level_ids =  {'c9ccb61ae9334416aba972f5d5089939', '3f8aea42eb324f08a16776671498dd1b'}       

# resulting adventure_ids is:
set() 
```

But if i first have a base (the level IDs) and update the lang_ids and tag_ids with the intersection with the base, and then take their union, it'd result in:

```{'c9ccb61ae9334416aba972f5d5089939', '3f8aea42eb324f08a16776671498dd1b'}```       

Does that make sense or am i missing anything?